core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -300,16 +332,49 @@ private void tryStart() {
             CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
             applyConfiguration(createCommand);
 
-            containerId = createCommand.exec().getId();
+            createCommand.getLabels().put(DockerClientFactory.TESTCONTAINERS_LABEL, "true");
 
-            connectToPortForwardingNetwork(createCommand.getNetworkMode());
+            boolean reused = false;
+            if (shouldBeReused) {
+                if (!canBeReused()) {
+                    throw new IllegalStateException("This container does not support reuse");
+                }
 
-            copyToFileContainerPathMap.forEach(this::copyFileToContainer);
+                if (TestcontainersConfiguration.getInstance().environmentSupportsReuse()) {
+                    String hash = hash(createCommand);
 
-            containerIsCreated(containerId);
+                    containerId = findContainerForReuse(hash).orElse(null);
 
-            logger().info("Starting container with ID: {}", containerId);
-            dockerClient.startContainerCmd(containerId).exec();
+                    if (containerId != null) {
+                        logger().info("Reusing container with ID: {} and hash: {}", containerId, hash);
+                        reused = true;
+                    } else {
+                        logger().debug("Can't find a reusable running container with hash: {}", hash);
+
+                        createCommand.getLabels().put(HASH_LABEL, hash);
+                    }
+                } else {
+                    logger().info("Reuse was requested but the environment does not support the reuse of containers");
```

Comment:
This warning will only appear if they use `withReuse(true)`, which means that they are aware of the feature already :)

Also, we expect this feature to be pretty well discoverable by anyone who cares about the speed of the tests (will also make sure that the docs explain it well), and I don't feel that we need the banner

Reply:
I just tried this feature for the first time, and I have to say it's completely non-obvious. I see the INFO log, but there is no hint about what I need to do to fix it.

Reply:
BTW the file name in the banner hint above appears to be wrong as well. I think it needs to be `~/.testcontainers.properties`. Windows users are probably going to be confused by that as well. Is it the only way to set that property?

Reply:
@aguibert FYI `updateGlobalConfig` should not be used for settings things that users should be opting-in themselves. I will even consider adding a whitelist or blacklist to the method.

Reply:
The reason why we discourage the in-memory use of global properties is because they SHOULD NOT be set by tests' code, but require a manual action by the user, or being used as a persistent storage (e.g. the selected strategy)

Features like this (reusable containers) is environment specific and cannot be assumed for everyone (think CI or shared environments), otherwise we would make it enabled by default.

Reply:
Maybe I am a bit to opinionated, but I always assume that Windows users, which are using Docker, should now that `~/` means $HOME directory, but I understand that this is not the perfect UX.

Reply:
Was just trying to debug this feature... it is very difficult to discover.

Additionally, it's setup such that it **forces** you to use `~/.testcontainers.properties` to configure it... why not allow it to also be set from the classpath `testcontainers.properties`?

By forcing it to be under `$HOME`, you're forcing end-users to setup all build servers, every developer workstation, etc to configure this file.  I don't see why you can't let the end-user opt-in to this feature with the classpath config file, this makes the most sense for enterprise development where consistency is king

Reply:
The intent of this feature (reusing containers across multiple test runs) was to only have it apply for developer environments. @bsideup specifically did not want it to be used for CI/build servers because container instances could pile up and bog down remote servers and easily go un-noticed.

That being said, I do think a more prominent INFO/WARN message for this situation would help users figure out how to turn this on in their local dev workstations.

Reply:
Thanks @aguibert for providing a good answer :)

So yeah, what Andy said :)

>  I do think a more prominent INFO/WARN message for this situation would help users figure out how to turn this on in their local dev workstations.

Yes, next iteration we will definitely improve the UX around it 👍

Reply:
What about if a user does check this in to the file on their classpath, to ignore it (as you are already doing), but also print out the explanation that @aguibert mentioned and that the setting needs to be set elsewhere.

(Perhaps that is what you meant by `improve the UX`)

Reply:
yes, we will definitely consider reporting environment-specific configuration set in the classpath file! Good suggestion 👍